roledb.go

```diff
@@ -293,19 +294,38 @@ func ReadRawGroups(importedGroups []RawAuthGroup) ([]AuthGroup, error) {
 // The 'cache' parameter is used to speed up subsequent calls to this function,
 // because this function tends to get used in loops.
 func GroupIDsToNames(groups []GroupIDU32, db RoleGroupDB, cache map[GroupIDU32]string) ([]string, error) {
-	names := make([]string, len(groups))
-	for i, gid := range groups {
+	// A group in 'groups' may not exist in the RoleGroupDB and we don't want to return a blank item in the string array.
+	// Instead we use append. It does mean a new slice variable gets created, but the overhead is minimal.
+	// Since the capacity of the slice is already maxed, no new underlying array is allocated during append.
+	names := make([]string, 0, len(groups))
+	for _, gid := range groups {
 		if cache[gid] != "" {
-			names[i] = cache[gid]
+			//names[i] = cache[gid]
+			names = append(names, cache[gid])
 		} else {
-			if group, err := db.GetByID(gid); err != nil {
+			// why is this not retrieved in one go? the below would result in 'n' nr of individual queries to db
+			// the below should only happen at most once per request
+			// if no new groups are added before the next request, cache should be up to date
+			// CAVEAT - In case of missing group, the cache will be retrieved EVERY time the group is referenced.
+			// Consider using a 'populated' flag
+			if groupsDB, err := db.GetGroups(); err != nil {
 				return nil, err
 			} else {
-				names[i] = group.Name
-				cache[gid] = group.Name
+				// re-populate groups cache
+				for _, giddb := range groupsDB {
+					if cache[giddb.ID] == "" {
+						cache[giddb.ID] = giddb.Name
+					}
+				}
+				// cycle through new groups, find in cache (based on new group id)
+				if cache[gid] != "" {
+					//names[i] = cache[gid]
+					names = append(names, cache[gid])
+				}
 			}
```

Comment:
Had a long rant here about how we should keep the old code, but based on your comment it looks like we were both wrong as to how this function is *used*.

The three places that we call this function in *our* auth service, we feed it an empty cache... which means that we will always end up hitting the db ALOT. My recommendation would be that the calling function should initialize the cache, to allow this function to remain smaller. But that would require us to create a function that does
```go
			cache := map[GroupIDU32]string{}

			if groupsDB, err := db.GetGroups(); err != nil {
				return nil, err
			}

			// populate groups cache
			for _, giddb := range groupsDB {
				if cache[giddb.ID] == "" {
						cache[giddb.ID] = giddb.Name
				}
			}
			return cache, nil
```

I would also compromise with checking the `len` of `cache` at the start of the function and then calling the new function to create a cache of the groups.